src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidEmissionProbability.kt

```diff
@@ -0,0 +1,188 @@
+     * return noReadCountProbability when there are no reads. Since the emission probability only needs to be
+     * proportional to the probability of the observations (read counts) given the state, returning 0.0 for all pairs works fine.
+     */
+    fun lnProbObsGivenState(state: Pair<SampleGamete, SampleGamete>, refrange: ReferenceRange): Double {
+
+        if (myCurrentRange != refrange) {
+            myCurrentRange = refrange
+            val readsForRange = readMap[refrange]
+            rangeHasReads = if (readsForRange == null) false else readsForRange.values.sum() > 0
+            if (rangeHasReads) assignRangeProbabilities()
+        }
+
+        if (!rangeHasReads) return noReadCountProbability
+
+        val haplotypePair = Pair(sampleToHaplotype[state.first], sampleToHaplotype[state.second])
+
+        //convert the pair of sample gametes to an UnorderedHaplotypePair
+        val unorderedHaplotypes = UnorderedHaplotypePair(haplotypePair)
+        return rangeLnProbabilities[unorderedHaplotypes] ?: defaultProbability
+    }
+
+    /**
+     * Generates a map of all possible SampleGamete order pairs to their emission probabilities. This is done
+     * once per reference range to avoid repeating the calculations.
+     */
+    private fun assignRangeProbabilities() {
+        sampleToHaplotype = createSampleToHaplotypeMap()
+
+        val haplotypesInRefrange = sampleToHaplotype.values.toSet().toList()
+
+        val readCounts = readMap[myCurrentRange] ?: mapOf()
+
+        require(readCounts.isNotEmpty()) {"No haplotypes in $myCurrentRange"}
+        val readSetCounts = readCounts.mapKeys { (haplist, _) -> haplist.toSet() }
+
+        val probabilityMap = mutableMapOf<UnorderedHaplotypePair, Double>()
+        for (ndx1 in haplotypesInRefrange.indices) {
+            for (ndx2 in ndx1 until haplotypesInRefrange.size) {
+                val haplotypePair = UnorderedHaplotypePair(Pair(haplotypesInRefrange[ndx1], haplotypesInRefrange[ndx2]))
+                probabilityMap[haplotypePair] = haplotypePairProbability(haplotypePair, readSetCounts)
+            }
+        }
+
+        //if there are any null haplotypes in this reference range add pairs for each (haplotype,null) and (null,null)
+        val anyNullHaplotypes = sampleGametesInGraph.any { sampleToHaplotype[it] == null }
+        if (anyNullHaplotypes) {
+            for (haplotype in haplotypesInRefrange) {
+                val haplotypePair = UnorderedHaplotypePair(Pair(haplotype, null))
+                probabilityMap[haplotypePair] = haplotypePairProbability(haplotypePair, readSetCounts)
+            }
+            val haplotypePair = UnorderedHaplotypePair(Pair(null, null))
+            probabilityMap[haplotypePair] = haplotypePairProbability(haplotypePair, readSetCounts)
+        }
+
+        //take the natural log of the probabilities
+        rangeLnProbabilities = probabilityMap.mapValues { (_, pr) -> ln(pr) }
```

Comment:
Could this be done on the fly above when you are assigning the initial probabilities?  Would save a loop through the states.